shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Value.hs

```diff
@@ -0,0 +1,407 @@
+import           Shelley.Spec.Ledger.Crypto
+import           Data.ByteString.Char8 (ByteString, pack) -- TODO is this the right Bytestring
+import           Shelley.Spec.Ledger.Scripts
+
+
+
+
+-- | Quantity
+newtype Quantity = Quantity Integer
+  deriving (Show, Eq, Generic, ToCBOR, FromCBOR, Ord, Integral, Num, Real, Enum, NoUnexpectedThunks)
+
+-- | Asset ID
+newtype AssetID = AssetID ByteString
+  deriving (Show, Eq, ToCBOR, FromCBOR, Ord, NoUnexpectedThunks)
+
+assetID :: AssetID -> ByteString
+assetID (AssetID aid) = aid
+
+-- | Policy ID
+newtype PolicyID crypto = PolicyID (ScriptHash crypto)
+  deriving (Show, Eq, ToCBOR, FromCBOR, Ord, NoUnexpectedThunks)
+
+policyID :: PolicyID crypto -> ScriptHash crypto
+policyID (PolicyID pid) = pid
+
+-- | A cryptocurrency value. This is a map from 'PolicyID's to a
+-- quantity of assets with this policy.
+--
+-- Operations on assets are usually implemented /pointwise/. That is,
+-- we apply the operation to the quantities for each asset in turn. So
+-- when we add two 'Value's the resulting 'Value' has, for each asset,
+-- the sum of the quantities of /that particular/ asset in the argument
+-- 'Value'. The effect of this is that the assets in the 'Value' are "independent",
+-- and are operated on separately.
+--
+-- Whenever we need to get the quantity of an asset in a 'Value' where there
+-- is no explicit quantity of that asset in the 'Value', then the quantity is
+-- taken to be zero.
+--
+-- See note [Assets] for more details.
+
+-- | Value type
+data Value crypto = Value
+  { val :: Map  (PolicyID crypto) (Map AssetID Quantity) }
+  deriving (Show, Generic)
+
+instance NoUnexpectedThunks (Value crypto)
+
+-- data ValueBSType = ValueBSType (Map ByteString (Map ByteString Quantity))
+--   deriving (Show, Eq, Generic)
+--
+-- instance NoUnexpectedThunks ValueBSType
+--
+-- -- | make a crypto-free Value type
+-- -- TODO this is a hack!
+-- toValBST :: Value crypto -> ValueBSType
+-- toValBST (Value v) = ValueBSType $ fromList $ fmap (\(cid, tkns) -> (pack $ show cid, tkns)) (toList v)
+
+-- | compact representation of Value
+data CompactValue crypto = AdaOnly Coin | MixValue (Value crypto)
+  deriving (Show, Eq, Generic)
+
+instance NoUnexpectedThunks (CompactValue crypto)
+
+
+-- get the quantities of the tokens of a value term
+getQs :: Value crypto -> [Quantity]
+getQs (Value v) = fmap snd (concat $ fmap toList (elems v))
+
+-- zero value
+zeroV :: Value crypto
+zeroV = Value empty
+
+-- | make a value out of a coin
+coinToValue :: Coin -> Value crypto
+coinToValue (Coin c) = Value $ singleton adaID (singleton adaToken (Quantity c))
+
+-- | make a value out of a coin
+getAdaAmount :: Value crypto -> Coin
+getAdaAmount (Value v) = Coin $ c
+  where
+    Quantity c = foldl (+) (Quantity 0) (getQs $ Value $ filterWithKey (\k _ -> k == adaID) v)
+
+-- | policy ID of Ada
+adaID :: PolicyID crypto
+adaID = PolicyID $ ScriptHash $ coerce ("" :: Hash ShortHash (Script crypto))
```

Comment:
This needs a note! What's going on here, why is it okay. This is an important permanent decision.